cupy/indexing/insert.py

```diff
@@ -107,6 +139,24 @@ def fill_diagonal(a, val, wrap=False):
             raise ValueError('All dimensions of input must be of equal length')
         step = 1 + numpy.cumprod(a.shape[:-1]).sum()
 
-    # Since the current cupy does not support a.flat,
-    # we use a.ravel() instead of a.flat
-    a.ravel()[:end:step] = val
+    if not isinstance(val, cupy.ndarray):
+        val = cupy.array(val, dtype=a.dtype)
+    else:
+        val = val.astype(a.dtype, copy=False)
+
```

Comment:
why not `val = cupy.asarray(val, dtype=a.dtype)`?

Reply:
Ah, it seems that this is valid in numpy numpy.fill_diagonal(a, [2,3]) ...

maybe just using `cupy.asarray(val)` if val is not a single value and then passing it by value in any other case.

Reply:
Yes, I followed NumPy here. And I fixed it to remove the branch!

Reply:
Thanks!!! 😄